auacm/app/static/js/controllers/ScoreboardController.js

```diff
@@ -61,38 +13,95 @@ app.controller('ScoreboardController', ['$scope', '$http', '$routeParams', '$win
+                            - thisTime;
+                    thisTime = Math.floor(new Date().getTime() / 1000);
+                    if ($scope.timeLeft <= 0) {
+                        $scope.active = false;
+                        $scope.timeLeft = 0;
+                        $interval.cancel(timer);
+                    }
+                }, 1000);
+            }
+            
         },
         function(error) {
             
         });
+}]);
 
-
-    // Open the socket connection
-    var socket = io.connect('http://' + $window.location.host + '/judge')
-    socket.on('connect', function() {
-        console.log('connected');
+var genScoreboard = function($scope) {
+    for (var i = 0; i < $scope.teams.length; i++) {
+        var team = $scope.teams[i];
+        var solved = 0;
+        var time = 0;
+        for (var problem in team.problemData) {
+            if (team.problemData[problem].status == 'correct') {
+                solved++;
+                time += team.problemData[problem].problemTime;
+            }
+        }
+        team.solved = solved;
+        team.time = time;
+    }
+    
+    $scope.teams.sort(function(a, b) {
+        if (a.solved != b.solved) {
+            return b.solved - a.solved;
+        } else {
+            return a.time - b.time;
+        }
     });
+    
+    var rank = 1;
+    var prevSolved = $scope.teams[0].solved;
+    var prevTime = $scope.teams[0].solved;
+    $scope.teams[0].rank = rank;
+    for (var i = 1; i < $scope.teams.length; i++) {
+        var team = $scope.teams[i];
+        if (team.solved < prevSolved) {
+            rank++;
+            team.rank = rank;
+        } else if (team.solved == prevSolved && team.time > prevTime) {
+            rank++;
+            team.rank = rank;
+        } else {
+            team.rank = rank;
+        }
+        prevSolved = team.solved;
+        prevTime = team.time;
+    }
+
+    // Can break if an $apply is already in progress, need to check first (???)
+    if (!$scope.$$phase) {
+        $scope.$digest();
+    }
+}
 
+var connectToSocket = function($window, $scope) {
+    // Open the socket connection
+    var socket = io.connect('http://' + $window.location.host + '/judge');
     // Perform live updates to the scoreboard
     var viewed = [];
     socket.on('status', function(event) {
-        if (viewed.indexOf(event.submissionId) > -1 || event.status == 'running') {
-            // The scoreboard doesn't care about the problem if it's not done or
-            // if we've already seen it.
+        if (viewed.indexOf(event.submissionId) > -1 || event.status == 'running'
+                || $scope.compProblems.indexOf(event.problemId) < 0
+                || event.submitTime > $scope.competition.startTime + $scope.competition.length) {
+            // The scoreboard ignores the problem for any of the following reasons:
+            // The submission has already been handled, the event's status is running,
```

Comment:
I know Kattis will update a scoreboard to show a problem is in the process of being judged. Do we want to consider this in future update? I don't know if ICPC does this.